New URL Bar Commands

papa-Leo opened the pull request:
This PR adds some new commands to the URL bar, as discussed in #10577:

#### Default commands
- [x] Reset Pinned Tab
- [x] Replace Pinned URL with Current
- [x] Duplicate Tab
- [x] Reopen Closed Tab
- [x] Bookmark Tab
- [ ] Move Current/Selected Tabs to Folder
- [ ] Move Current/Selected Tabs to Space

#### Advanced preference commands
- [x] Create New Workspace
- [x] Delete Current Workspace
- [x] Unsplit Tabs
- [x] Unsplit Current Tab
- [x] Split View Vertical
- [x] Split View Horizontal
- [x] Split View Grid
- [x] Copy Current URL Markdown
- [x] Search Bookmarks
- [ ] Swap Split Tabs
- [ ] Open in Container

💬 More suggestions are welcome, please put them in the comments.

papa-Leo committed:
Added 'Create New Workspace' command

papa-Leo committed:
Added 'Delete Current Workspace' command

mr-cheffy commented:
I'll convert this into a draft since I see not all of them are done. Please ping me or undraft it once its ready for review, thanks!

papa-Leo committed:
Added 'Unsplit Tabs' command

12th-devs commented:
Is the warning for deleting the current workspace still present? I would not want someone to be typing something, have that prioritized then have there main workspace deleted with no hope

papa-Leo commented:
> Is the warning for deleting the current workspace still present? I would not want someone to be typing something, have that prioritized then have there main workspace deleted with no hope

I've implemented this with the `cmd_zenCtxDeleteWorkspace`, so yes the confirmation is still present. 👍

12th-devs commented:
> > Is the warning for deleting the current workspace still present? I would not want someone to be typing something, have that prioritized then have there main workspace deleted with no hope
> 
> I've implemented this with the `cmd_zenCtxDeleteWorkspace`, so yes the confirmation is still present. 👍

Awesome!

(Sorry for not looking at the pr)

mr-cheffy commented:
To be honest, I dont think deleting or creating spaces is something that should be considered "quick action"...

papa-Leo commented:
> To be honest, I don't think deleting or creating spaces is something that should be considered "quick action"...

I’ve implemented them because they’re useful for my personal workflow, but it’s true that they’re a bit niche.

Obviously the final decision is up to you, but I propose that they’re disabled by default, and can be enabled with an advanced preference flag. This keeps things simple but makes it available for those who need it.

I’d love to hear thoughts on this

12th-devs commented:
This file might be of help to you: https://github.com/BibekBhusal0/zen-custom-js/blob/main/command-palette/all-commands.js

papa-Leo committed:
Added 'Unsplit Current Tab' command

papa-Leo commented:
> This file might be of help to you: https://github.com/BibekBhusal0/zen-custom-js/blob/main/command-palette/all-commands.js

This is great! Thanks for sharing 😀

papa-Leo committed:
Added 'Reset Pinned Tab' command

papa-Leo committed:
Added 'Replace Pinned URL with Current' command

papa-Leo committed:
Added 'Split Selected Tabs' command

papa-Leo committed:
Added 'Duplicate Tab' command

papa-Leo committed:
Added 'Copy Current URL Markdown' command

papa-Leo committed:
Added 'Split View Vertical', 'Split View Horizontal', 'Split View Grid'
commands. Removed 'Split Selected Tabs' command, redundant

NexusOnePlus commented:
I think it could also be useful to have commands for working with containers.

My suggestions would be:
 - Use Container (Select a container and reload the current tab in it)
 - Open in Container (Choose a container and open the query/URL in a new tab within that container)

mr-cheffy commented:
I also wouldn't like having `Copy Current URL Markdown` and all of the split related commands here as they would add clutter when the actions aren't really that necessary or alternatives already exist

papa-Leo commented:
> I also wouldn't like having `Copy Current URL Markdown` and all of the split related commands here as they would add clutter when the actions aren't really that necessary or alternatives already exist

@mr-cheffy what do you think about my [previously mentioned](https://github.com/zen-browser/desktop/pull/10603#issuecomment-3344200436) suggestion that these extra commands are disabled by default, but can be enabled with an advanced preference flag for power users?
These extra commands would make a huge difference to my workflow, and I'm sure that's true for others.

papa-Leo commented:
I see the "command palette" feature as a way to make less frequently used commands more accessible - especially those with complicated keyboard shortcuts or no shortcuts at all. That’s why I was focusing on these specific commands.

But I understand if this approach doesn’t align with your goal for the feature. This is your project, so I'll go with whatever you choose.
I appreciate your guidance

papa-Leo commented:
These are the commands I recommend moving to advanced preferences:
- Split View Vertical
- Split View Horizontal
- Split View Grid
- Unsplit All Tabs
- Create New Workspace
- Delete Current Workspace
- Copy Current URL Markdown
- Move Tab to Folder
- Move Tab to Workspace
- Container related commands as mentioned [by @NexusOnePlus previously](https://github.com/zen-browser/desktop/pull/10603#issuecomment-3348418250)

Do you have an upper limit in mind to the number of commands that are enabled by default?

willrossignol commented:
I found myself looking for a "Bookmark current tab" action while using the new URL bar, that'd be sweet!

mr-cheffy commented:
I agree @papa-Leo, the pref should be named `zen.urlbar.suggestions.quick-actions.advanced`.  

> Move Tab to Folder
> Move Tab to Workspace

These however can stay as they are more common actions

papa-Leo committed:
Added 'Reopen Closed Tab' command

papa-Leo committed:
Added 'Bookmark Tab', 'Search Bookmarks' commands

papa-Leo committed:
Add quick actions advanced mode behind preference flag

papa-Leo committed:
Merge branch 'zen-browser:dev' into new-urlbar-commands

mr-cheffy commented:
@papa-Leo actions are heuristic, meaning they need to be fetched *fast*, and getting the values of preferences are *very* slow. 

I would rather this pref be considered a startup pref. Meaning it'll only be read once.

e.g.
```js
var actions = [
  { normal action },
  { normal action },
  ...(advanced pref getter) ? [ advanced prefs ] : []
]
```

papa-Leo committed:
Refactored fetching of advanced quick actions pref

papa-Leo committed:
Merge branch 'zen-browser:dev' into new-urlbar-commands

mr-cheffy commented:
Let me know when its ready for a review!

UnnaturalTwilight commented:
Could a command to open the saved passwords page (about:logins)
there are probably other about: pages that having a command for would be useful

mr-cheffy committed:
Merge branch 'dev' into new-urlbar-commands

drakoooo commented:
any chance a command to search opened tabs could also be implemented ? thanks for your work anyways !

mr-cheffy commented:
Search for tabs already exists. Write % and the space

mr-cheffy commented:
Please mark the PR as ready for review if you'd like for me to start taking a look. We can merge the basic commands now for twilight testing and the more complex ones can be separate PRs

sualk01 commented:
It would be nice to have a command to directly prompt the currently selected ai bot, instead of having to open the site pane and then to enter there the command.

for example: "\command\ summarize the current page"

LemonPotion commented:
Is this PR still alive?
